swift/ql/lib/codeql/swift/regex/Regex.qll

```diff
@@ -8,23 +8,56 @@ private import codeql.swift.dataflow.DataFlow
 private import internal.ParseRegex
 private import internal.RegexTracking
 
+/**
+ * A data flow node whose value may flow to a position where it is interpreted
+ * as a part of a regular expression. For example the string literal
+ * `"(a|b).*"` in:
+ * ```
+ * Regex("(a|b).*").firstMatch(in: myString)
+ * ```
+ */
+abstract class RegexPatternSource extends DataFlow::Node {
+  /**
+   * Gets a node where the pattern of this node is parsed as a part of
+   * a regular expression.
+   */
+  abstract DataFlow::Node getAParse();
+
+  /**
+   * Gets the root term of the regular expression parsed from this pattern.
+   */
+  abstract RegExpTerm getRegExpTerm();
+}
+
+/**
+ * For each `RegexPatternSource` data flow node, the corresponding `Expr` is
+ * a `Regex`. This is a simple wrapper to make that happen.
+ */
+private class RegexFromRegexPatternSource extends RegExp {
+  RegexPatternSource node;
+
+  RegexFromRegexPatternSource() { this = node.asExpr() }
+}
+
 /**
  * A string literal that is used as a regular expression. For example
  * the string literal `"(a|b).*"` in:
  * ```
  * Regex("(a|b).*").firstMatch(in: myString)
  * ```
  */
-private class ParsedStringRegex extends RegExp, StringLiteralExpr {
+private class ParsedStringRegex extends RegexPatternSource {
+  StringLiteralExpr expr;
   DataFlow::Node use;
 
-  ParsedStringRegex() { StringLiteralUseFlow::flow(DataFlow::exprNode(this), use) }
+  ParsedStringRegex() {
+    expr = this.asExpr() and
+    StringLiteralUseFlow::flow(this, use)
+  }
 
-  /**
-   * Gets a dataflow node where this string literal is used as a regular
-   * expression.
-   */
-  DataFlow::Node getUse() { result = use }
+  override DataFlow::Node getAParse() { result = use }
+
+  override RegExpTerm getRegExpTerm() { result.getRegExp() = this.asExpr() }
```

Comment:
```suggestion
  override RegExpTerm getRegExpTerm() { result.getRegExp() = expr }
```

Reply:
Thanks - that's why I'm getting warnings from QL-for-QL as well.